### PR TITLE
Enhanced RDoc for Array

### PR DESCRIPTION
**Pull request by BurdetteLamar**
Methods:
- cycle
- permutation
- combination

### CODE REVIEW

**Comment by BurdetteLamar** [comment]
@marcandre, I very much appreciate your reviews, which have identified a number of errors and omissions.

As for the structure/ordering of the doc for the methods:  I've written thus for nearly 200 methods in ENV, Hash, and now Array, so I think it best to finish up these last 15 methods the same way.

I do think that the structure of the methods doc needs to be discussed.  I've considered opening an Issue, but there may be a better way of capturing everyone's views.  Topics I'm interested in include:
- (Nearly) always beginning with a one-line sentence that gives a sense of the use of the method.
- For multi-signature (correct term?) methods, a sensible way to re-state each signature as it's discussed.  I've used horizontal rules to help in this regard, but I'd love to re-state the actual signature as well.

There are probably other discussion-worthy issues.

If/when such things are settled, I'll be happy to revise the doc for the 200 methods, as well as others I've written over at JSON and CSV.
I'll get to your other comments shortly.  And thanks for your interest.

File: array.c [hunk 1]
```diff
@@ -7492,30 +7514,86 @@ rb_ary_permutation_size(VALUE ary, VALUE args, VALUE eobj)
...
- *     ary.permutation(n)                     -> Enumerator
+ *    array.permutation {|element| ... } -> self
+ *    array.permutation(n) {|element| ... } -> self
+ *    array.permutation -> new_enumerator
+ *    array.permutation(n) -> new_enumerator
  *
- * When invoked with a block, yield all permutations of length +n+ of the
- * elements of the array, then return the array itself.
+ *  Calls the block, if given, with permutations of elements of +self+; returns +self+.
+ *  The order of permutations is indeterminate.
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply array.c:7523]: This is consistent with many already-merged doc patches.

File: array.c [hunk 2]
```diff
@@ -7492,30 +7514,86 @@ rb_ary_permutation_size(VALUE ary, VALUE args, VALUE eobj)
...
  *
- * If +n+ is not specified, yield all permutations of all elements.
+ *  Argument +n+, if given, must be an
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply array.c:7525]: This is consistent with many already-merged doc patches.

File: array.c [hunk 3]
```diff
@@ -7492,30 +7514,86 @@ rb_ary_permutation_size(VALUE ary, VALUE args, VALUE eobj)
...
+ *  When +n+ is out of range (negative or larger than <tt>self.size</tt>),
+ *  does not call the block:
+ *    a = [0, 1, 2]
+ *    a.permutation(-1) {|permutation| fail 'Cannot happen' }
+ *    a.permutation(4) {|permutation| fail 'Cannot happen' }
+ *
+ *  ---
+ *
+ *  When a block given but no argument,
+ *  behaves the same as <tt>a.permutation(a.size)</tt>:
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply array.c:7574]: I don't see that.  All examples above this have an argument?

File: array.c [hunk 4]
```diff
@@ -7329,24 +7329,46 @@ rb_ary_cycle_size(VALUE self, VALUE args, VALUE eobj)
...
+ *
+ *  ---
+ *
+ *  When no block is given, returns a new \Enumerator:
  *
- *     a = ["a", "b", "c"]
- *     a.cycle {|x| puts x}       # print, a, b, c, a, b, c,.. forever.
- *     a.cycle(2) {|x| puts x}    # print, a, b, c, a, b, c.
+ *    [0, 1].cycle(2) # => #<Enumerator: [0, 1]:cycle(2)>
+ *    [0, 1].cycle # => # => #<Enumerator: [0, 1]:cycle>
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply array.c:7364]: Added, but many similar passages lack examples.

[the remark this answers is not in the document]
Comment (marcandre) [reply array.c:7364]: > Added, but many similar passages lack examples.

I fail to see the point you are trying to make.

File: array.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (marcandre) [reply array.c:7525]: I'm giving up.

**Comment by marcandre** [comment]
Clearly, I am failing in making myself understood so I will stop commenting on anything else than factual errors.

File: array.c [hunk 5]
```diff
@@ -7492,30 +7514,86 @@ rb_ary_permutation_size(VALUE ary, VALUE args, VALUE eobj)
...
- *     ary.permutation(n) {|p| block}         -> ary
- *     ary.permutation(n)                     -> Enumerator
+ *    array.permutation {|element| ... } -> self
+ *    array.permutation(n) {|element| ... } -> self
+ *    array.permutation -> new_enumerator
+ *    array.permutation(n) -> new_enumerator
  *
- * When invoked with a block, yield all permutations of length +n+ of the
- * elements of the array, then return the array itself.
+ *  Calls the block, if given, with permutations of elements of +self+; returns +self+.
```
[the remark this answers is not in the document]
Comment (BurdetteLamar) [reply array.c:7522]: Fixed.

**Comment by BurdetteLamar** [comment]
@jeremyevans, if you still approve this PR, would you kindly merge it?

**Comment by BurdetteLamar** [comment]
@jeremyevans, I totally agree.  As I've said, I think we need broad agreement about how the doc should be.  When we have that, I'll be glad to rework what I've done this year.

But first, I'd like to finish this pass (~15 methods to go), remaining consistent with the 200+, however deficient.  Consistency counts.

**Pull request merged**